shell/platform/fuchsia/flutter/vulkan_surface.cc

```diff
@@ -226,57 +219,105 @@ bool VulkanSurface::CreateFences() {
+  image_constraints = fuchsia::sysmem::ImageFormatConstraints();
+  image_constraints.min_coded_width = size.width();
+  image_constraints.min_coded_height = size.height();
+  image_constraints.max_coded_width = size.width();
+  image_constraints.max_coded_height = size.height();
+  image_constraints.min_bytes_per_row = 0;
+  image_constraints.pixel_format.type = kSysmemPixelFormat;
+  image_constraints.color_spaces_count = 1;
+  image_constraints.color_space[0].type = fuchsia::sysmem::ColorSpaceType::SRGB;
+
+  status = buffer_collection->SetConstraints(true, constraints);
+  LOG_AND_RETURN(status != ZX_OK, "Failed to set constraints");
+
+  VkBufferCollectionCreateInfoFUCHSIA import_info;
+  import_info.collectionToken = vulkan_token.Unbind().TakeChannel().release();
+  if (VK_CALL_LOG_ERROR(vulkan_provider_.vk().CreateBufferCollectionFUCHSIA(
+          vulkan_provider_.vk_device(), &import_info, nullptr, &collection_)) !=
+      VK_SUCCESS) {
     return false;
   }
 
+  VulkanImage vulkan_image;
+  LOG_AND_RETURN(!CreateVulkanImage(vulkan_provider_, size, &vulkan_image),
+                 "Failed to create VkImage");
+
+  status = buffer_collection->Close();
+  LOG_AND_RETURN(status != ZX_OK, "Failed to close collection");
+
   vulkan_image_ = std::move(vulkan_image);
-  const VkMemoryRequirements& memory_reqs =
+  const VkMemoryRequirements& memory_requirements =
       vulkan_image_.vk_memory_requirements;
-  const VkImageCreateInfo& image_create_info =
-      vulkan_image_.vk_image_create_info;
-
-  uint32_t memory_type = 0;
-  for (; memory_type < 32; memory_type++) {
-    if ((memory_reqs.memoryTypeBits & (1 << memory_type))) {
-      break;
-    }
-  }
+  VkImageCreateInfo& image_create_info = vulkan_image_.vk_image_create_info;
 
-  VkMemoryDedicatedAllocateInfo dedicated_allocate_info = {
-      .sType = VK_STRUCTURE_TYPE_MEMORY_DEDICATED_ALLOCATE_INFO,
+  VkImportMemoryBufferCollectionFUCHSIA import_memory_info = {
+      .sType = VK_STRUCTURE_TYPE_IMPORT_MEMORY_BUFFER_COLLECTION_FUCHSIA,
       .pNext = nullptr,
-      .image = vulkan_image_.vk_image,
-      .buffer = VK_NULL_HANDLE};
-  VkExportMemoryAllocateInfoKHR export_allocate_info = {
-      .sType = VK_STRUCTURE_TYPE_EXPORT_MEMORY_ALLOCATE_INFO_KHR,
-      .pNext = &dedicated_allocate_info,
-      .handleTypes =
-          VK_EXTERNAL_MEMORY_HANDLE_TYPE_TEMP_ZIRCON_VMO_BIT_FUCHSIA};
-
-  const VkMemoryAllocateInfo alloc_info = {
+      .collection = collection_,
+      .index = 0,
+  };
+  auto memory_type = memory_requirements.memoryTypeBits;
+  VkMemoryAllocateInfo allocation_info = {
       .sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
-      .pNext = &export_allocate_info,
-      .allocationSize = memory_reqs.size,
-      .memoryTypeIndex = memory_type,
+      .pNext = &import_memory_info,
+      .allocationSize = memory_requirements.size,
+      .memoryTypeIndex =
+          memory_type == 0 ? 32 : (uint32_t)__builtin_ctz(memory_type),
```

Comment:
- why is the "memory_type == 0" check needed? if needed, why not "0 ? 0 : ... "
- nit: static_cast instead of c-style cast

Reply:
I don't think `memoryTypeBits` are guaranteed not to be `0` and `__builtin_ctz` has UB for `0`. I'll change this to an `FML_DCHECK`.

Reply:
If memoryTypeBits is 0 then something is very wrong and we can't allocate memory for the image so DCHECK makes sense.